src/jdk.compiler/share/classes/com/sun/tools/javac/code/Type.java

```diff
@@ -1161,6 +1182,54 @@ public boolean isReference() {
             return true;
         }
 
+        @Override
+        public boolean isValue() {
+            return tsym != null && tsym.isValue();
+        }
+
+        @Override
+        public boolean isReferenceProjection() {
+            return tsym != null && tsym.isReferenceProjection();
+        }
+
+        @Override
+        public Type valueProjection() {
+            if (!isReferenceProjection() || projection !=  null)
```

Comment:
Question here - since you have two symbols, getting a reference projection and value projection shouldn't be as simple as doing `tsym.XYZprojection().type` ? Why do we need to create new types here?

Reply:
FTR, this is to preserve any parameterizations. The value projection of V.ref<String> is V.val<String> and not V.val<T>